File: app/src/main/java/com/bitcoin/merchant/app/util/AmountUtil.java
package com.bitcoin.merchant.app.util;

import android.content.Context;
import android.util.Log;

import com.bitcoin.merchant.app.currency.CountryCurrency;
import com.bitcoin.merchant.app.currency.CurrencyHelper;

import java.text.NumberFormat;
import java.util.Currency;

public class AmountUtil {
    public static final String TAG = "AmountUtil";
    public static final String DEFAULT_CURRENCY_BCH = "BCH";
    private final Context context;

    public AmountUtil(Context context) {
        this.context = context;
    }

    public String formatFiat(double amountFiat) {
        String currency = AppUtil.getCurrency(context);
        String country = AppUtil.getCountry(context);
        String locale = AppUtil.getLocale(context);
        CountryCurrency cc = CurrencyHelper.getInstance(context).getCountryCurrency(currency, country, locale);
        String fiat = null;
        try {
            NumberFormat formatter = NumberFormat.getCurrencyInstance(cc.countryLocales.getLocale());
            Currency instance = Currency.getInstance(currency);
            formatter.setCurrency(instance);
            formatter.setMaximumFractionDigits(instance.getDefaultFractionDigits());
            fiat = formatter.format(amountFiat);
        } catch (Exception e) {
            Log.d(TAG, "Locale not supported for " + currency + " failed to format to fiat: " + amountFiat);
        }
        if (fiat != null) {
            String currencySign = "\u00a4";
            fiat = fiat.replace(currencySign, currency);
        } else {
            fiat = currency + " " + MonetaryUtil.getInstance().getFiatDecimalFormat().format(amountFiat);
        }
        return fiat;
    }

    public String formatBch(double amountBch) {
        return MonetaryUtil.getInstance().getBchDecimalFormat().format(amountBch) + " " + DEFAULT_CURRENCY_BCH;
    }
}

File: app/src/main/java/com/bitcoin/merchant/app/util/AmountUtil.kt
package com.bitcoin.merchant.app.util

import android.content.Context
import android.util.Log
import com.bitcoin.merchant.app.model.Analytics
import java.text.NumberFormat
import java.util.*

class AmountUtil(private val context: Context) {
    fun formatFiat(amountFiat: Double): String {
        val ccl = Settings.getCountryCurrencyLocale(context)
        return try {
            val formatter = NumberFormat.getCurrencyInstance(ccl.locale)
            formatter.currency = Currency.getInstance(ccl.currency)
            formatter.maximumFractionDigits = ccl.decimals
            formatter.format(amountFiat).replace(CURRENCY_SIGN, ccl.currency)
        } catch (e: Exception) {
            Analytics.error_format_currency.sendError(e, ccl.locale.country, ccl.currency, ccl.locale.displayName)
            Log.d(TAG, "Locale not supported for $ccl.currency failed to format to fiat: $amountFiat")
            ccl.currency + " " + MonetaryUtil.instance.fiatDecimalFormat.format(amountFiat)
        }
    }

    fun satsToBch(satoshis: Long): String {
        return formatBch(satoshis / 100000000.0, false)
    }

    fun formatBch(amountBch: Double): String {
        return formatBch(amountBch, false)
    }

    fun formatBch(amountBch: Double, unit: Boolean): String {
        var string = MonetaryUtil.instance.bchDecimalFormat.format(amountBch)
        if(unit) string += " $DEFAULT_CURRENCY_BCH"
        return string
    }

    companion object {
        const val TAG = "BCR-AmountUtil"
        const val DEFAULT_CURRENCY_BCH = "BCH"
        const val CURRENCY_SIGN = "\u00a4"
    }
}